<a href="https://colab.research.google.com/github/siriwat-heart/DWDM21/blob/main/Project_%E0%B8%A5%E0%B9%88%E0%B8%B2%E0%B8%AA%E0%B8%B8%E0%B8%94.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
import pandas as pd #ส่วนของการ Import
import os

In [6]:
from google.colab import drive 
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [7]:
path = '/content/drive/My Drive/DWDM2021_Project' #กำหนด path ของโฟลเดอร์ที่ Data อยู่

# ที่มาของข้อมูลสถานีรับส่งผู้โดยสารการรถไฟเเห่งประเทศไทย
https://data.go.th/dataset/station

In [8]:
data_st = pd.read_csv(os.path.join(path,'data_station.csv')) #กำหนดตัวแปร data_st เป็นข้อมูลของสถานีรถไฟแต่ละสถานี
data_st #เรียกดูตารางของสถานีรถไฟแต่ละสถานี

id  station_code           name  ... active giveway dual_track
0      1          1001        กรุงเทพ  ...      1       1          1
1      2          1002          ยมราช  ...      1       1          1
2      3          1004         สามเสน  ...      1       1          1
3      4          1006     ประดิพัทธ์  ...      1       1          1
4      5          1007  ชุมทางบางซื่อ  ...      1       1          1
..   ...           ...            ...  ...    ...     ...        ...
677  678          5025       บางโทรัด  ...      1       0          0
678  679          5026      บ้านกาหลง  ...      1       0          0
679  680          5027     บ้านนาขวาง  ...      1       0          0
680  681          5028      บ้านนาโคก  ...      1       0          0
681  682          5029       เขตเมือง  ...      1       0          0

[682 rows x 17 columns]

# ที่มาของข้อมูลขบวนรถไฟสำหรับรับ-ส่งผู้โดยสาร
https://data.go.th/dataset/train

In [9]:
data_t = pd.read_csv(os.path.join(path,'data_train.csv')) #กำหนดตัวแปร data_t เป็นข้อมูลของรถไฟ
data_t #เรียกดูตารางข้อมูลรถไฟ

Zone  goback   Num  ...            NameStEng            NameStEng_L  Service
0       1       1   0.0  ...              Bangkok       Ban Plu Ta Luang    False
1       1       1   1.0  ...              Bangkok      Bang Sue Junction     True
2       1       1   1.0  ...             Makkasan              Hua Takhe    False
3       1       1   2.0  ...              Bangkok      Bang Sue Junction     True
4       1       1   4.0  ...              Bangkok       Ban Plu Ta Luang    False
..    ...     ...   ...  ...                  ...                    ...      ...
229     5       2  26.0  ...              Nam Tok              Thon Buri    False
230     5       2  27.0  ...  Prachuap Khiri Khan              Thon Buri    False
231     5       2  28.0  ...            Lang Suan              Thon Buri    False
232     5       2  29.0  ...              Nam Tok              Thon Buri    False
233     5       2  30.0  ...              Nam Tok  Nong Pla Duk Junction    False

[234 rows x 12 columns]

# ที่มาของข้อมูล railstation
https://datagov.mot.go.th/dataset/train-station

In [10]:
data_rst = pd.read_csv(os.path.join(path,'railstation_new.csv')) #กำหนดตัวแปร data_rst ให้เป็นรางรถไฟ
data_rst #เรียกดูตารางของรางรถไฟ

OBJECTID                      NAMET  ...           X          Y
0           1               บ้านเกาะปริง  ...   99.644994   8.213274
1           2                    ห้วยยอด  ...   99.637987   7.785456
2           3            โรงเรียนการบิน   ...   99.956660  14.088246
3           4                บ้านปลักแรด  ...  100.725785  14.530267
4           5                    เด่นชัย  ...  100.046225  17.980781
..        ...                        ...  ...         ...        ...
677       678  ศูนย์ราชการจังหวัดสระแก้ว  ...  102.138802  13.799641
678       679                    ทุ่งทอง  ...   99.638000  13.980168
679       680                ตลาดหนองคาย  ...  102.725416  17.878784
680       681               แกรนด์ คาแนล  ...  100.608267  13.941516
681       682                    กรุงเทพ  ...  100.516732  13.744026

[682 rows x 8 columns]

# ที่มาของข้อมูลชื่อตำบล อำเภอ และจังหวัด
https://www.bot.or.th/Thai/Statistics/DataManagementSystem/TempClose/FI_FM1/Code/DocLib_StandardCodeMapping/Location_Postal%20Code.xls

In [11]:
data_location = pd.read_csv(os.path.join(path,'Location_Postal_Code.csv')) ## นำเข้าข้อมูลชื่อตำบล อำเภอ และจังหวัด
data_location ##เรียกดูข้อมูลชื่อตำบล อำเภอ และจังหวัด

Location Code   จังหวัด   เขต/อำเภอ        แขวง/ตำบล  รหัสไปรษณีย์
0            100101   กรุงเทพ      พระนคร  พระบรมมหาราชวัง         10200
1            100102   กรุงเทพ      พระนคร   วังบูรพาภิรมย์         10200
2            100103   กรุงเทพ      พระนคร       วัดราชบพิธ         10200
3            100104   กรุงเทพ      พระนคร      สำราญราษฎร์         10200
4            100105   กรุงเทพ      พระนคร   ศาลเจ้าพ่อเสือ         10200
...             ...       ...         ...              ...           ...
7421         961203  นราธิวาส       จะแนะ         ผดุงมาตร         96220
7422         961204  นราธิวาส       จะแนะ        ช้างเผือก         96220
7423         961301  นราธิวาส  เจาะไอร้อง              จวบ         96130
7424         961302  นราธิวาส  เจาะไอร้อง            บูกิต         96130
7425         961303  นราธิวาส  เจาะไอร้อง       มะรือโบออก         96130

[7426 rows x 5 columns]

# ที่มาของข้อมูลสถานการณ์การท่องเที่ยวในประเทศ รายจังหวัด ปี 2563
https://mots.go.th/more_news_new.php?cid=594

In [12]:
data_tourist = pd.read_csv(os.path.join(path,'สถานการณ์การท่องเที่ยวในประเทศ-รายจังหวัด-ปี-2563.csv')) ## นำเข้าข้อมูลสถานการณ์การท่องเที่ยวในประเทศ รายจังหวัด ปี 2563
data_tourist ##เรียกดูข้อมูล

ภาค          จังหวัด  จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)
0                 ภาคกลาง          กรุงเทพ                       23617914.0
1                 ภาคกลาง           ลพบุรี                        1659868.0
2                 ภาคกลาง  พระนครศรีอยุธยา                        3560866.0
3                 ภาคกลาง          สระบุรี                        1673819.0
4                 ภาคกลาง           ชัยนาท                         394900.0
..                    ...              ...                              ...
72  ภาคตะวันออกเฉียงเหนือ      อุบลราชธานี                        1537178.0
73  ภาคตะวันออกเฉียงเหนือ           สกลนคร                         856974.0
74  ภาคตะวันออกเฉียงเหนือ            ยโสธร                         303461.0
75  ภาคตะวันออกเฉียงเหนือ       อำนาจเจริญ                         153692.0
76  ภาคตะวันออกเฉียงเหนือ      หนองบัวลำภู                         191384.0

[77 rows x 3 columns]

# ดูว่าข้อมูลมี data missing

In [13]:
data_st.isnull() ##ดูค่า Missing ในตาราง data_st

id  station_code   name  en_name  ...   long  active  giveway  dual_track
0    False         False  False    False  ...  False   False    False       False
1    False         False  False    False  ...  False   False    False       False
2    False         False  False    False  ...  False   False    False       False
3    False         False  False    False  ...  False   False    False       False
4    False         False  False    False  ...  False   False    False       False
..     ...           ...    ...      ...  ...    ...     ...      ...         ...
677  False         False  False    False  ...  False   False    False       False
678  False         False  False    False  ...  False   False    False       False
679  False         False  False    False  ...  False   False    False       False
680  False         False  False    False  ...  False   False    False       False
681  False         False  False    False  ...  False   False    False       False

[682 rows x 17 columns]

In [14]:
data_t.isnull() ##ดูค่า Missing ในตาราง data_t

Zone  goback    Num  IdTrain  ...  NameSt_L  NameStEng  NameStEng_L  Service
0    False   False  False    False  ...     False      False        False    False
1    False   False  False    False  ...     False      False        False    False
2    False   False  False    False  ...     False      False        False    False
3    False   False  False    False  ...     False      False        False    False
4    False   False  False    False  ...     False      False        False    False
..     ...     ...    ...      ...  ...       ...        ...          ...      ...
229  False   False  False    False  ...     False      False        False    False
230  False   False  False    False  ...     False      False        False    False
231  False   False  False    False  ...     False      False        False    False
232  False   False  False    False  ...     False      False        False    False
233  False   False  False    False  ...     False      False        False    False

[234 rows x 12 columns]

In [15]:
data_rst.isnull() ##ดูค่า Missing ในตาราง data_rst

OBJECTID  NAMET  NAMEE  ProvName  AmpName  TamName      X      Y
0       False  False  False     False    False     True  False  False
1       False  False  False     False    False    False  False  False
2       False  False  False     False    False    False  False  False
3       False  False  False     False    False    False  False  False
4       False  False  False     False    False    False  False  False
..        ...    ...    ...       ...      ...      ...    ...    ...
677     False  False  False     False    False    False  False  False
678     False  False  False     False    False    False  False  False
679     False  False  False     False    False    False  False  False
680     False  False  False     False    False    False  False  False
681     False  False  False     False    False    False  False  False

[682 rows x 8 columns]

In [16]:
data_location.isnull() ##ดูค่า Missing ในตาราง data_location

Location Code  จังหวัด  เขต/อำเภอ  แขวง/ตำบล  รหัสไปรษณีย์
0             False    False      False      False         False
1             False    False      False      False         False
2             False    False      False      False         False
3             False    False      False      False         False
4             False    False      False      False         False
...             ...      ...        ...        ...           ...
7421          False    False      False      False         False
7422          False    False      False      False         False
7423          False    False      False      False         False
7424          False    False      False      False         False
7425          False    False      False      False         False

[7426 rows x 5 columns]

In [17]:
data_tourist.isnull() ##ดูค่า Missing ในตาราง data_prov

ภาค  จังหวัด  จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)
0   False    False                            False
1   False    False                            False
2   False    False                            False
3   False    False                            False
4   False    False                            False
..    ...      ...                              ...
72  False    False                            False
73  False    False                            False
74  False    False                            False
75  False    False                            False
76  False    False                            False

[77 rows x 3 columns]

# สรุปข้อมูลเป็นรายคอลลัมม์ว่ามี missing

In [18]:
data_st.isnull().any()  ##ดูค่า Missing รายคอลัมน์ของตาราง data_st

id                 False
station_code       False
name               False
en_name            False
th_short           False
en_short            True
chname              True
controldivision    False
exact_km           False
exact_distance     False
km                 False
class              False
lat                False
long               False
active             False
giveway            False
dual_track         False
dtype: bool

In [19]:
data_t.isnull().any() ##ดูค่า Missing รายคอลัมน์ของตาราง data_t

Zone           False
goback         False
Num            False
IdTrain        False
NameFee        False
StStart        False
NameSt         False
StFinal        False
NameSt_L       False
NameStEng      False
NameStEng_L    False
Service        False
dtype: bool

In [21]:
data_rst.isnull().any() ##ดูค่า Missing รายคอลัมน์ของตาราง data_rst

OBJECTID    False
NAMET       False
NAMEE       False
ProvName    False
AmpName     False
TamName      True
X           False
Y           False
dtype: bool

In [22]:
data_location.isnull().any()  ##ดูค่า Missing รายคอลัมน์ของตาราง data_location

Location Code    False
จังหวัด          False
เขต/อำเภอ         True
แขวง/ตำบล        False
รหัสไปรษณีย์     False
dtype: bool

In [23]:
data_tourist.isnull().any()  ##ดูค่า Missing รายคอลัมน์ของตาราง data_prov

ภาค                                False
จังหวัด                            False
จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)    False
dtype: bool

# Drop missing value
ลบค่าที่เป็น Missing

In [24]:
data_st = data_st.dropna() ##ลบค่า Missing ในตาราง data_st
data_st ##เรียกดูตาราง data_st

id  station_code           name  ... active giveway dual_track
0      1          1001        กรุงเทพ  ...      1       1          1
1      2          1002          ยมราช  ...      1       1          1
2      3          1004         สามเสน  ...      1       1          1
3      4          1006     ประดิพัทธ์  ...      1       1          1
4      5          1007  ชุมทางบางซื่อ  ...      1       1          1
..   ...           ...            ...  ...    ...     ...        ...
677  678          5025       บางโทรัด  ...      1       0          0
678  679          5026      บ้านกาหลง  ...      1       0          0
679  680          5027     บ้านนาขวาง  ...      1       0          0
680  681          5028      บ้านนาโคก  ...      1       0          0
681  682          5029       เขตเมือง  ...      1       0          0

[679 rows x 17 columns]

In [25]:
data_rst = data_rst.dropna() ##ลบค่า Missing ในตาราง data_rst
data_rst ##เรียกดูตาราง data_rst

OBJECTID                      NAMET  ...           X          Y
1           2                    ห้วยยอด  ...   99.637987   7.785456
2           3            โรงเรียนการบิน   ...   99.956660  14.088246
3           4                บ้านปลักแรด  ...  100.725785  14.530267
4           5                    เด่นชัย  ...  100.046225  17.980781
5           6                บ้านกกกว้าว  ...  100.442303  15.141322
..        ...                        ...  ...         ...        ...
677       678  ศูนย์ราชการจังหวัดสระแก้ว  ...  102.138802  13.799641
678       679                    ทุ่งทอง  ...   99.638000  13.980168
679       680                ตลาดหนองคาย  ...  102.725416  17.878784
680       681               แกรนด์ คาแนล  ...  100.608267  13.941516
681       682                    กรุงเทพ  ...  100.516732  13.744026

[681 rows x 8 columns]

In [26]:
data_location = data_location.dropna() ##ลบค่า Missing ในตาราง data_location
data_location ##เรียกดูตาราง data_location

Location Code   จังหวัด   เขต/อำเภอ        แขวง/ตำบล  รหัสไปรษณีย์
0            100101   กรุงเทพ      พระนคร  พระบรมมหาราชวัง         10200
1            100102   กรุงเทพ      พระนคร   วังบูรพาภิรมย์         10200
2            100103   กรุงเทพ      พระนคร       วัดราชบพิธ         10200
3            100104   กรุงเทพ      พระนคร      สำราญราษฎร์         10200
4            100105   กรุงเทพ      พระนคร   ศาลเจ้าพ่อเสือ         10200
...             ...       ...         ...              ...           ...
7421         961203  นราธิวาส       จะแนะ         ผดุงมาตร         96220
7422         961204  นราธิวาส       จะแนะ        ช้างเผือก         96220
7423         961301  นราธิวาส  เจาะไอร้อง              จวบ         96130
7424         961302  นราธิวาส  เจาะไอร้อง            บูกิต         96130
7425         961303  นราธิวาส  เจาะไอร้อง       มะรือโบออก         96130

[7425 rows x 5 columns]

# ตรวจสอบเมื่อลบออกไปแล้วตรวจสอบว่ายังไม่ข้อมูลที่หายไปหรือไม่

In [27]:
data_st.isnull().any() ##ดูค่า Missing รายคอลัมน์ของตาราง data_st หลังการลบค่า Missing

id                 False
station_code       False
name               False
en_name            False
th_short           False
en_short           False
chname             False
controldivision    False
exact_km           False
exact_distance     False
km                 False
class              False
lat                False
long               False
active             False
giveway            False
dual_track         False
dtype: bool

In [28]:
data_rst.isnull().any() ##ดูค่า Missing รายคอลัมน์ของตาราง data_rst หลังการลบค่า Missing

OBJECTID    False
NAMET       False
NAMEE       False
ProvName    False
AmpName     False
TamName     False
X           False
Y           False
dtype: bool

In [29]:
data_location.isnull().any() ##ดูค่า Missing รายคอลัมน์ของตาราง data_location หลังการลบค่า Missing

Location Code    False
จังหวัด          False
เขต/อำเภอ        False
แขวง/ตำบล        False
รหัสไปรษณีย์     False
dtype: bool

# Percent of missing data from dropna

Percent of missing data from dropna
เปอร์เซ็นต์ข้อมูลที่สูญเสีย

In [30]:
##ตาราง data_st
st_b = 682
st_a = 679
print(f'before = {st_b} \nafter = {st_a}\nwhich is {100*((st_b-st_a)/st_b)}')

before = 682 
after = 679
which is 0.43988269794721413


In [31]:
##ตาราง data_rst
rst_b = 682
rst_a = 681
print(f'before = {rst_b} \nafter = {rst_a}\nwhich is {100*((rst_b-rst_a)/rst_b)}')

before = 682 
after = 681
which is 0.1466275659824047


In [32]:
##ตาราง data_location
loca_b = 7426
loca_a = 7425
print(f'before = {loca_b} \nafter = {loca_a}\nwhich is {100*((loca_b-loca_a)/loca_b)}')

before = 7426 
after = 7425
which is 0.013466199838405602


# รวมตารางหาจังหวัดต้นทางต้นทาง

In [33]:
table_st1 = data_t.merge(data_rst,how='left',left_on='NameSt',right_on='NAMET') #ทำการเชื่อมตาราง 2 ตาราง คือ data_t และ data_rst
table_st1 ##เรียกดูตาราง

Zone  goback   Num  ...              TamName           X          Y
0       1       1   0.0  ...          RONG MUEANG  100.516732  13.744026
1       1       1   1.0  ...          RONG MUEANG  100.516732  13.744026
2       1       1   1.0  ...             MAKKASAN  100.547316  13.753736
3       1       1   2.0  ...          RONG MUEANG  100.516732  13.744026
4       1       1   4.0  ...          RONG MUEANG  100.516732  13.744026
..    ...     ...   ...  ...                  ...         ...        ...
229     5       2  26.0  ...                        99.067848  14.232451
230     5       2  27.0  ...  PRACHUAP KHIRI KHAN   99.793905  11.807757
231     5       2  28.0  ...            LANG SUAN   99.076502   9.949263
232     5       2  29.0  ...                        99.067848  14.232451
233     5       2  30.0  ...                        99.067848  14.232451

[234 rows x 20 columns]

In [34]:
table_st1.isnull().any() ##ดูค่า Missing ในตารางราย Column

Zone           False
goback         False
Num            False
IdTrain        False
NameFee        False
StStart        False
NameSt         False
StFinal        False
NameSt_L       False
NameStEng      False
NameStEng_L    False
Service        False
OBJECTID        True
NAMET           True
NAMEE           True
ProvName        True
AmpName         True
TamName         True
X               True
Y               True
dtype: bool

In [35]:
table_st1 = table_st1.dropna() ##ลบค่า Missing ในตาราง
table_st1 ##เรียกดูตาราง

Zone  goback   Num  ...              TamName           X          Y
0       1       1   0.0  ...          RONG MUEANG  100.516732  13.744026
1       1       1   1.0  ...          RONG MUEANG  100.516732  13.744026
2       1       1   1.0  ...             MAKKASAN  100.547316  13.753736
3       1       1   2.0  ...          RONG MUEANG  100.516732  13.744026
4       1       1   4.0  ...          RONG MUEANG  100.516732  13.744026
..    ...     ...   ...  ...                  ...         ...        ...
229     5       2  26.0  ...                        99.067848  14.232451
230     5       2  27.0  ...  PRACHUAP KHIRI KHAN   99.793905  11.807757
231     5       2  28.0  ...            LANG SUAN   99.076502   9.949263
232     5       2  29.0  ...                        99.067848  14.232451
233     5       2  30.0  ...                        99.067848  14.232451

[225 rows x 20 columns]

In [36]:
table_st1.isnull().any() ##ดูค่า Missing ในตารางราย Column หลังลบค่า Missing

Zone           False
goback         False
Num            False
IdTrain        False
NameFee        False
StStart        False
NameSt         False
StFinal        False
NameSt_L       False
NameStEng      False
NameStEng_L    False
Service        False
OBJECTID       False
NAMET          False
NAMEE          False
ProvName       False
AmpName        False
TamName        False
X              False
Y              False
dtype: bool

In [37]:
##ตาราง table_st1
table_st1_b = 234
table_st1_a = 225
print(f'before = {table_st1_b} \nafter = {table_st1_a}\nwhich is {100*((table_st1_b-table_st1_a)/table_st1_b)}')

before = 234 
after = 225
which is 3.8461538461538463


In [38]:
table_st2 = table_st1.merge(data_location,how='left',left_on='AmpName',right_on='เขต/อำเภอ') #ทำการเชื่อมตาราง 2 ตาราง คือ table_st1 และ data_location
table_st2 ##เรียกดูตาราง

Zone  goback   Num  IdTrain  ...    จังหวัด  เขต/อำเภอ  แขวง/ตำบล  รหัสไปรษณีย์
0        1       1   0.0      997  ...    กรุงเทพ    ปทุมวัน   รองเมือง       10330.0
1        1       1   0.0      997  ...    กรุงเทพ    ปทุมวัน    วังใหม่       10330.0
2        1       1   0.0      997  ...    กรุงเทพ    ปทุมวัน    ปทุมวัน       10330.0
3        1       1   0.0      997  ...    กรุงเทพ    ปทุมวัน    ลุมพินี       10330.0
4        1       1   1.0     1049  ...    กรุงเทพ    ปทุมวัน   รองเมือง       10330.0
...    ...     ...   ...      ...  ...        ...        ...        ...           ...
2395     5       2  30.0      486  ...  กาญจนบุรี     ไทรโยค      สิงห์       71150.0
2396     5       2  30.0      486  ...  กาญจนบุรี     ไทรโยค     ไทรโยค       71150.0
2397     5       2  30.0      486  ...  กาญจนบุรี     ไทรโยค  วังกระแจะ       71150.0
2398     5       2  30.0      486  ...  กาญจนบุรี     ไทรโยค    ศรีมงคล       71150.0
2399     5       2  30.0      486  ...  กาญจนบุรี     ไทรโยค    บ้องตี้       71150.0

[2400 rows x 25 columns]

In [39]:
table_st2.isnull().any() ##ดูค่า Missing ในตาราง

Zone             False
goback           False
Num              False
IdTrain          False
NameFee          False
StStart          False
NameSt           False
StFinal          False
NameSt_L         False
NameStEng        False
NameStEng_L      False
Service          False
OBJECTID         False
NAMET            False
NAMEE            False
ProvName         False
AmpName          False
TamName          False
X                False
Y                False
Location Code     True
จังหวัด           True
เขต/อำเภอ         True
แขวง/ตำบล         True
รหัสไปรษณีย์      True
dtype: bool

In [40]:
table_st2 = table_st2.dropna() #กำจัดค่า Missing
table_st2 #เรียกดูตาราง

Zone  goback   Num  IdTrain  ...    จังหวัด  เขต/อำเภอ  แขวง/ตำบล  รหัสไปรษณีย์
0        1       1   0.0      997  ...    กรุงเทพ    ปทุมวัน   รองเมือง       10330.0
1        1       1   0.0      997  ...    กรุงเทพ    ปทุมวัน    วังใหม่       10330.0
2        1       1   0.0      997  ...    กรุงเทพ    ปทุมวัน    ปทุมวัน       10330.0
3        1       1   0.0      997  ...    กรุงเทพ    ปทุมวัน    ลุมพินี       10330.0
4        1       1   1.0     1049  ...    กรุงเทพ    ปทุมวัน   รองเมือง       10330.0
...    ...     ...   ...      ...  ...        ...        ...        ...           ...
2395     5       2  30.0      486  ...  กาญจนบุรี     ไทรโยค      สิงห์       71150.0
2396     5       2  30.0      486  ...  กาญจนบุรี     ไทรโยค     ไทรโยค       71150.0
2397     5       2  30.0      486  ...  กาญจนบุรี     ไทรโยค  วังกระแจะ       71150.0
2398     5       2  30.0      486  ...  กาญจนบุรี     ไทรโยค    ศรีมงคล       71150.0
2399     5       2  30.0      486  ...  กาญจนบุรี     ไทรโยค    บ้องตี้       71150.0

[2398 rows x 25 columns]

In [41]:
table_st2.isnull().any() #ดูค่า Missing ในตารางราย Column

Zone             False
goback           False
Num              False
IdTrain          False
NameFee          False
StStart          False
NameSt           False
StFinal          False
NameSt_L         False
NameStEng        False
NameStEng_L      False
Service          False
OBJECTID         False
NAMET            False
NAMEE            False
ProvName         False
AmpName          False
TamName          False
X                False
Y                False
Location Code    False
จังหวัด          False
เขต/อำเภอ        False
แขวง/ตำบล        False
รหัสไปรษณีย์     False
dtype: bool

In [42]:
##ตาราง table_st2
table_st2_b = 2400
table_st2_a = 2398
print(f'before = {table_st2_b} \nafter = {table_st2_a}\nwhich is {100*((table_st2_b-table_st2_a)/table_st2_b)}')

before = 2400 
after = 2398
which is 0.08333333333333334


จากการคำนวณจะเห็นได้ว่าตาราง table_st2 มีข้อมูลที่สูญเสีย 0.08 %

In [43]:
table_st3 = table_st2.merge(data_tourist,how='left',left_on='จังหวัด',right_on='จังหวัด') #ทำการเชื่อมตาราง 2 ตาราง คือ table_st2 และ data_tourist
table_st3 #เรียกดูตาราง

Zone  goback  ...         ภาค  จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)
0        1       1  ...     ภาคกลาง                       23617914.0
1        1       1  ...     ภาคกลาง                       23617914.0
2        1       1  ...     ภาคกลาง                       23617914.0
3        1       1  ...     ภาคกลาง                       23617914.0
4        1       1  ...     ภาคกลาง                       23617914.0
...    ...     ...  ...         ...                              ...
2393     5       2  ...  ภาคตะวันตก                        6068266.0
2394     5       2  ...  ภาคตะวันตก                        6068266.0
2395     5       2  ...  ภาคตะวันตก                        6068266.0
2396     5       2  ...  ภาคตะวันตก                        6068266.0
2397     5       2  ...  ภาคตะวันตก                        6068266.0

[2398 rows x 27 columns]

In [44]:
table_st3.isnull().any() #ดูค่า Missing ราย Column

Zone                               False
goback                             False
Num                                False
IdTrain                            False
NameFee                            False
StStart                            False
NameSt                             False
StFinal                            False
NameSt_L                           False
NameStEng                          False
NameStEng_L                        False
Service                            False
OBJECTID                           False
NAMET                              False
NAMEE                              False
ProvName                           False
AmpName                            False
TamName                            False
X                                  False
Y                                  False
Location Code                      False
จังหวัด                            False
เขต/อำเภอ                          False
แขวง/ตำบล                          False
รหัสไปรษณีย์    

# เลือกเฉพาะ column ที่ต้องใช้

In [45]:
table_st = table_st3[['Zone','NameFee','StStart','NameSt','StFinal','NameSt_L','X','Y','จังหวัด','ภาค','จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)']] ##เลือกคอลัมน์ที่ต้องการ
table_st ##แสดงข้อมูลของตารางที่เลือก

Zone   NameFee  ...         ภาค จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)
0        1      เร็ว  ...     ภาคกลาง                      23617914.0
1        1      เร็ว  ...     ภาคกลาง                      23617914.0
2        1      เร็ว  ...     ภาคกลาง                      23617914.0
3        1      เร็ว  ...     ภาคกลาง                      23617914.0
4        1    ธรรมดา  ...     ภาคกลาง                      23617914.0
...    ...       ...  ...         ...                             ...
2393     5  ท้องถิ่น  ...  ภาคตะวันตก                       6068266.0
2394     5  ท้องถิ่น  ...  ภาคตะวันตก                       6068266.0
2395     5  ท้องถิ่น  ...  ภาคตะวันตก                       6068266.0
2396     5  ท้องถิ่น  ...  ภาคตะวันตก                       6068266.0
2397     5  ท้องถิ่น  ...  ภาคตะวันตก                       6068266.0

[2398 rows x 11 columns]

In [46]:
table_st.isnull().any() #ดูค่า Missing ราย Column

Zone                               False
NameFee                            False
StStart                            False
NameSt                             False
StFinal                            False
NameSt_L                           False
X                                  False
Y                                  False
จังหวัด                            False
ภาค                                False
จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)    False
dtype: bool

# **รวมตารางหาจังหวัดปลายทาง**

In [47]:
table_l1 = data_st.merge(data_rst,how='right',left_on='name',right_on='NAMET') #ทำการเชื่อมตาราง 2 ตาราง คือ data_st และ data_rst
table_l1 #เรียกดูตาราง

id  station_code          name  ...      TamName           X          Y
0    565.0        4281.0       ห้วยยอด  ...                99.637987   7.785456
1      NaN           NaN           NaN  ...                99.956660  14.088246
2     28.0        1038.0   บ้านปลักแรด  ...               100.725785  14.530267
3    102.0        1164.0       เด่นชัย  ...     DEN CHAI  100.046225  17.980781
4     43.0        1062.0   บ้านกกกว้าว  ...               100.442303  15.141322
..     ...           ...           ...  ...          ...         ...        ...
676    NaN           NaN           NaN  ...               102.138802  13.799641
677  414.0        4052.0       ทุ่งทอง  ...                99.638000  13.980168
678  297.0        2305.0   ตลาดหนองคาย  ...               102.725416  17.878784
679  138.0        1233.0  แกรนด์ คาแนล  ...               100.608267  13.941516
680    1.0        1001.0       กรุงเทพ  ...  RONG MUEANG  100.516732  13.744026

[681 rows x 25 columns]

In [48]:
table_l1.isnull().any() ##ดูค่า Missing ราย Column

id                  True
station_code        True
name                True
en_name             True
th_short            True
en_short            True
chname              True
controldivision     True
exact_km            True
exact_distance      True
km                  True
class               True
lat                 True
long                True
active              True
giveway             True
dual_track          True
OBJECTID           False
NAMET              False
NAMEE              False
ProvName           False
AmpName            False
TamName            False
X                  False
Y                  False
dtype: bool

In [49]:
table_l1 = table_l1.dropna() #ลบค่า Missing
table_l1 #เรียกดูตาราง

id  station_code           name  ...      TamName           X          Y
0    565.0        4281.0        ห้วยยอด  ...                99.637987   7.785456
2     28.0        1038.0    บ้านปลักแรด  ...               100.725785  14.530267
3    102.0        1164.0        เด่นชัย  ...     DEN CHAI  100.046225  17.980781
4     43.0        1062.0    บ้านกกกว้าว  ...               100.442303  15.141322
5    351.0        3078.0        หนองสัง  ...               101.825379  13.980567
..     ...           ...            ...  ...          ...         ...        ...
675  375.0        3124.0  ตลาดน้ำ 4 ภาค  ...               100.909693  12.869605
677  414.0        4052.0        ทุ่งทอง  ...                99.638000  13.980168
678  297.0        2305.0    ตลาดหนองคาย  ...               102.725416  17.878784
679  138.0        1233.0   แกรนด์ คาแนล  ...               100.608267  13.941516
680    1.0        1001.0        กรุงเทพ  ...  RONG MUEANG  100.516732  13.744026

[658 rows x 25 columns]

In [50]:
table_l1.isnull().any() ##ดูค่า Missing ราย Column

id                 False
station_code       False
name               False
en_name            False
th_short           False
en_short           False
chname             False
controldivision    False
exact_km           False
exact_distance     False
km                 False
class              False
lat                False
long               False
active             False
giveway            False
dual_track         False
OBJECTID           False
NAMET              False
NAMEE              False
ProvName           False
AmpName            False
TamName            False
X                  False
Y                  False
dtype: bool

In [51]:
##ตาราง table_l1
table_l1_b = 681
table_l1_a = 658
print(f'before = {table_l1_b} \nafter = {table_l1_a}\nwhich is {100*((table_l1_b-table_l1_a)/table_l1_b)}')

before = 681 
after = 658
which is 3.3773861967694567


จากการคำนวณจะเห็นได้ว่าตาราง table_l1 มีข้อมูลที่สูญเสีย 3.37 %

In [52]:
table_l2 = table_l1.merge(data_location,how='left',left_on='AmpName',right_on='เขต/อำเภอ') #ทำการเชื่อมตาราง 2 ตาราง คือ table_l1 และ data_st
table_l2 ##เรียกดูตาราง

id  station_code          name  ... เขต/อำเภอ     แขวง/ตำบล รหัสไปรษณีย์
0     565.0        4281.0       ห้วยยอด  ...   ห้วยยอด       ห้วยยอด      92130.0
1     565.0        4281.0       ห้วยยอด  ...   ห้วยยอด  หนองช้างแล่น      92130.0
2     565.0        4281.0       ห้วยยอด  ...   ห้วยยอด         บางดี      92210.0
3     565.0        4281.0       ห้วยยอด  ...   ห้วยยอด       บางกุ้ง      92210.0
4     565.0        4281.0       ห้วยยอด  ...   ห้วยยอด        เขากอบ      92130.0
...     ...           ...           ...  ...       ...           ...          ...
7060  138.0        1233.0  แกรนด์ คาแนล  ...  ดอนเมือง       สนามบิน      10210.0
7061    1.0        1001.0       กรุงเทพ  ...   ปทุมวัน      รองเมือง      10330.0
7062    1.0        1001.0       กรุงเทพ  ...   ปทุมวัน       วังใหม่      10330.0
7063    1.0        1001.0       กรุงเทพ  ...   ปทุมวัน       ปทุมวัน      10330.0
7064    1.0        1001.0       กรุงเทพ  ...   ปทุมวัน       ลุมพินี      10330.0

[7065 rows x 30 columns]

In [53]:
table_l2.isnull().any() #ดูค่า Missing ราย column

id                 False
station_code       False
name               False
en_name            False
th_short           False
en_short           False
chname             False
controldivision    False
exact_km           False
exact_distance     False
km                 False
class              False
lat                False
long               False
active             False
giveway            False
dual_track         False
OBJECTID           False
NAMET              False
NAMEE              False
ProvName           False
AmpName            False
TamName            False
X                  False
Y                  False
Location Code       True
จังหวัด             True
เขต/อำเภอ           True
แขวง/ตำบล           True
รหัสไปรษณีย์        True
dtype: bool

In [54]:
table_l2 = table_l2.dropna() #ลบค่า Missing
table_l2 #เรียกดูตาราง

id  station_code          name  ... เขต/อำเภอ     แขวง/ตำบล รหัสไปรษณีย์
0     565.0        4281.0       ห้วยยอด  ...   ห้วยยอด       ห้วยยอด      92130.0
1     565.0        4281.0       ห้วยยอด  ...   ห้วยยอด  หนองช้างแล่น      92130.0
2     565.0        4281.0       ห้วยยอด  ...   ห้วยยอด         บางดี      92210.0
3     565.0        4281.0       ห้วยยอด  ...   ห้วยยอด       บางกุ้ง      92210.0
4     565.0        4281.0       ห้วยยอด  ...   ห้วยยอด        เขากอบ      92130.0
...     ...           ...           ...  ...       ...           ...          ...
7060  138.0        1233.0  แกรนด์ คาแนล  ...  ดอนเมือง       สนามบิน      10210.0
7061    1.0        1001.0       กรุงเทพ  ...   ปทุมวัน      รองเมือง      10330.0
7062    1.0        1001.0       กรุงเทพ  ...   ปทุมวัน       วังใหม่      10330.0
7063    1.0        1001.0       กรุงเทพ  ...   ปทุมวัน       ปทุมวัน      10330.0
7064    1.0        1001.0       กรุงเทพ  ...   ปทุมวัน       ลุมพินี      10330.0

[7035 rows x 30 columns]

In [55]:
table_l2.isnull().any() #ดูค่า Missing ราย column

id                 False
station_code       False
name               False
en_name            False
th_short           False
en_short           False
chname             False
controldivision    False
exact_km           False
exact_distance     False
km                 False
class              False
lat                False
long               False
active             False
giveway            False
dual_track         False
OBJECTID           False
NAMET              False
NAMEE              False
ProvName           False
AmpName            False
TamName            False
X                  False
Y                  False
Location Code      False
จังหวัด            False
เขต/อำเภอ          False
แขวง/ตำบล          False
รหัสไปรษณีย์       False
dtype: bool

In [56]:
##ตาราง table_l2
table_l2_b = 7065
table_l2_a = 7035
print(f'before = {table_l2_b} \nafter = {table_l2_a}\nwhich is {100*((table_l2_b-table_l2_a)/table_l2_b)}')

before = 7065 
after = 7035
which is 0.42462845010615713


In [57]:
table_l3 = table_l2.merge(data_tourist,how='left',left_on='จังหวัด',right_on='จังหวัด') #ทำการเชื่อมตาราง 2 ตาราง คือ table_l3 และ data_tourist
table_l3 #เรียกดูตาราง

id  station_code  ...      ภาค จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)
0     565.0        4281.0  ...   ภาคใต้                        648092.0
1     565.0        4281.0  ...   ภาคใต้                        648092.0
2     565.0        4281.0  ...   ภาคใต้                        648092.0
3     565.0        4281.0  ...   ภาคใต้                        648092.0
4     565.0        4281.0  ...   ภาคใต้                        648092.0
...     ...           ...  ...      ...                             ...
7030  138.0        1233.0  ...  ภาคกลาง                      23617914.0
7031    1.0        1001.0  ...  ภาคกลาง                      23617914.0
7032    1.0        1001.0  ...  ภาคกลาง                      23617914.0
7033    1.0        1001.0  ...  ภาคกลาง                      23617914.0
7034    1.0        1001.0  ...  ภาคกลาง                      23617914.0

[7035 rows x 32 columns]

In [58]:
table_l3.isnull().any() #ดูค่า Missing ราย column

id                                 False
station_code                       False
name                               False
en_name                            False
th_short                           False
en_short                           False
chname                             False
controldivision                    False
exact_km                           False
exact_distance                     False
km                                 False
class                              False
lat                                False
long                               False
active                             False
giveway                            False
dual_track                         False
OBJECTID                           False
NAMET                              False
NAMEE                              False
ProvName                           False
AmpName                            False
TamName                            False
X                                  False
Y               

# เลือกเฉพาะ column ที่ต้องใช้

In [59]:
table_l = table_l3[['station_code','name','exact_km','km','class','X','Y','จังหวัด','ภาค','จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)']] ##เลือกคอลัมน์ที่ต้องการ
table_l ##แสดงข้อมูลของตารางที่เลือก

station_code          name  ...      ภาค  จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)
0           4281.0       ห้วยยอด  ...   ภาคใต้                         648092.0
1           4281.0       ห้วยยอด  ...   ภาคใต้                         648092.0
2           4281.0       ห้วยยอด  ...   ภาคใต้                         648092.0
3           4281.0       ห้วยยอด  ...   ภาคใต้                         648092.0
4           4281.0       ห้วยยอด  ...   ภาคใต้                         648092.0
...            ...           ...  ...      ...                              ...
7030        1233.0  แกรนด์ คาแนล  ...  ภาคกลาง                       23617914.0
7031        1001.0       กรุงเทพ  ...  ภาคกลาง                       23617914.0
7032        1001.0       กรุงเทพ  ...  ภาคกลาง                       23617914.0
7033        1001.0       กรุงเทพ  ...  ภาคกลาง                       23617914.0
7034        1001.0       กรุงเทพ  ...  ภาคกลาง                       23617914.0

[7035 rows x 10 columns]

# **รวมตารางต้นทางและปลายทาง**

In [60]:
table_mix = table_st.merge(table_l,how='left',left_on='StFinal',right_on='station_code') #ทำการเชื่อมตาราง 2 ตาราง คือ table_st และ table_l
table_mix #เรียกดูตาราง

Zone   NameFee  ...        ภาค_y จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)_y
0         1      เร็ว  ...  ภาคตะวันออก                         6964743.0
1         1      เร็ว  ...  ภาคตะวันออก                         6964743.0
2         1      เร็ว  ...  ภาคตะวันออก                         6964743.0
3         1      เร็ว  ...  ภาคตะวันออก                         6964743.0
4         1      เร็ว  ...  ภาคตะวันออก                         6964743.0
...     ...       ...  ...          ...                               ...
21971     5  ท้องถิ่น  ...   ภาคตะวันตก                         1430360.0
21972     5  ท้องถิ่น  ...   ภาคตะวันตก                         1430360.0
21973     5  ท้องถิ่น  ...   ภาคตะวันตก                         1430360.0
21974     5  ท้องถิ่น  ...   ภาคตะวันตก                         1430360.0
21975     5  ท้องถิ่น  ...   ภาคตะวันตก                         1430360.0

[21976 rows x 21 columns]

In [61]:
table_mix.isnull().any() #ดูค่า Missing ราย column

Zone                                 False
NameFee                              False
StStart                              False
NameSt                               False
StFinal                              False
NameSt_L                             False
X_x                                  False
Y_x                                  False
จังหวัด_x                            False
ภาค_x                                False
จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)_x    False
station_code                          True
name                                  True
exact_km                              True
km                                    True
class                                 True
X_y                                   True
Y_y                                   True
จังหวัด_y                             True
ภาค_y                                 True
จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)_y     True
dtype: bool

In [62]:
table_mix = table_mix.dropna() #ลบค่า Missing
table_mix #เรียกดูตาราง

Zone   NameFee  ...        ภาค_y จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)_y
0         1      เร็ว  ...  ภาคตะวันออก                         6964743.0
1         1      เร็ว  ...  ภาคตะวันออก                         6964743.0
2         1      เร็ว  ...  ภาคตะวันออก                         6964743.0
3         1      เร็ว  ...  ภาคตะวันออก                         6964743.0
4         1      เร็ว  ...  ภาคตะวันออก                         6964743.0
...     ...       ...  ...          ...                               ...
21971     5  ท้องถิ่น  ...   ภาคตะวันตก                         1430360.0
21972     5  ท้องถิ่น  ...   ภาคตะวันตก                         1430360.0
21973     5  ท้องถิ่น  ...   ภาคตะวันตก                         1430360.0
21974     5  ท้องถิ่น  ...   ภาคตะวันตก                         1430360.0
21975     5  ท้องถิ่น  ...   ภาคตะวันตก                         1430360.0

[21870 rows x 21 columns]

In [63]:
table_mix.isnull().any() #ดูค่า Missing ราย column

Zone                                 False
NameFee                              False
StStart                              False
NameSt                               False
StFinal                              False
NameSt_L                             False
X_x                                  False
Y_x                                  False
จังหวัด_x                            False
ภาค_x                                False
จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)_x    False
station_code                         False
name                                 False
exact_km                             False
km                                   False
class                                False
X_y                                  False
Y_y                                  False
จังหวัด_y                            False
ภาค_y                                False
จำนวนผู้เยี่ยมเยือนทั้งหมด (คน)_y    False
dtype: bool

In [64]:
##ตาราง table_mix
table_mix_b = 21976
table_mix_a = 21870
print(f'before = {table_mix_b} \nafter = {table_mix_a}\nwhich is {100*((table_mix_b-table_mix_a)/table_mix_b)}')

before = 21976 
after = 21870
which is 0.48234437568256283


จากการคำนวณจะเห็นได้ว่าตาราง table_mix มีข้อมูลที่สูญเสีย 0.48 %